### PR TITLE
Add Microsoft Authentication

### PR DESCRIPTION
**Pull request by rtm516**
This is all functional, may need some more testing and some cleanup of the HTTP requests would be appreciated as I didn't really know what to do with them.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
     <groupId>com.github.steveice10</groupId>
     <artifactId>mcauthlib</artifactId>
-    <version>1.4-SNAPSHOT</version>
+    <version>1.5-SNAPSHOT</version>
```
Comment (Steveice10) [inline pom.xml:8]: I would recommend leaving it at 1.4-SNAPSHOT since there have been no changes yet for 1.4; this would be the first.

**Comment by Steveice10** [comment]
Did a look-over of the code and left a comment. Not sure if this is ready for final review yet since it's still marked as a draft.

**Comment by rtm516** [comment]
I have 1 thing left to add which is token auth for Microsoft in case we have already got one.

**Comment by tycrek** [comment]
Any progress on this PR? Would really appreciate being able to use it in one of my own projects.

**Comment by rtm516** [comment]
> Any progress on this PR? Would really appreciate being able to use it in one of my own projects.

Other than it not being super clean it does work, we have been using in in Geyser for the last month and a bit and have not had reports of major issues with it.

**Comment by tycrek** [comment]
> 
> Other than it not being super clean it does work, we have been using in in Geyser for the last month and a bit and have not had reports of major issues with it.

Awesome! If I wanted to test it experimentally, would it be enough to build the source from your fork? Or do you have a build on maven?

**Comment by rtm516** [comment]
> Or do you have a build on

We currently use a jitpack build, its 1 commit behind but latest should work the same
https://github.com/GeyserMC/Geyser/blob/master/connector/pom.xml#L242-L246

**Comment by tycrek** [comment]
> We currently use a jitpack build, its 1 commit behind but latest should work the same
> [https://github.com/GeyserMC/Geyser/blob/master/connector/pom.xml#L242-L246](https://github.com/GeyserMC/Geyser/blob/master/connector/pom.xml?rgh-link-date=2021-03-09T17%3A53%3A06Z#L242-L246)

I was able to get it added to my project and it appears to sign in fine, but it's unable to connect to servers (in my case, 1.12.2 Paper):

```
[09/03/2021 13:32:27] [AUTH] [SUCCESS] Logged in: TheTrueMEG (5ef0f4ad-534e-4b32-90b1-933c5b90f24d)
[09/03/2021 13:32:27] [GHOST] [INFO] Connecting...
[09/03/2021 13:32:27] [GHOST] [SUCCESS] Connected to [removed public IP address]:25565!
[09/03/2021 13:32:27] [GHOST] [INFO] Connected.
[09/03/2021 13:32:27] [CORE] [INFO] Waiting before running modules...
[09/03/2021 13:32:27] [GHOST] [INFO] Disconnecting from server...
[09/03/2021 13:32:27] [GHOST] [INFO] Disconnected.
[09/03/2021 13:32:27] [GHOST] [INFO] Disconnect reason: java.lang.NoSuchMethodError: com.github.steveice10.mc.auth.service.SessionService.<init>(Ljava/net/Proxy;)V
[09/03/2021 13:32:27] [GHOST] [WARN] Thread Thread-15 threw Exception null // this error is unrelated to AuthLib
[09/03/2021 13:32:28] [CORE] [INFO] Ghost has been stopped!
[09/03/2021 13:32:28] [CORE] [INFO] Reconnecting: yes
[09/03/2021 13:32:28] [CORE] [INFO] Waiting 31 seconds before attempting reconnect...
```

The full trace:

```
java.lang.NoSuchMethodError: com.github.steveice10.mc.auth.service.SessionService.<init>(Ljava/net/Proxy;)V
	at com.github.steveice10.mc.protocol.ClientListener.packetReceived(ClientListener.java:55)
	at com.github.steveice10.packetlib.event.session.PacketReceivedEvent.call(PacketReceivedEvent.java:51)
	at com.github.steveice10.packetlib.tcp.TcpSession.callEvent(TcpSession.java:143)
	at com.github.steveice10.packetlib.tcp.TcpPacketCodec.decode(TcpPacketCodec.java:48)
	at io.netty.handler.codec.ByteToMessageCodec$1.decode(ByteToMessageCodec.java:42)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:428)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
	at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
	at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
	at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:647)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:582)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:499)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:461)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```

I do understand that this is still experimental. I hope these logs are helpful in getting MS auth implemented :) if this issue is not fixable with the current codebase I am fine with waiting until it can be fully utilized, I'm just eager to get MS auth into my current projet

**Comment by rtm516** [comment]
@tycrek You are passing a java.net.Proxy to the SessionService constructor which doesn't take that as an argument. This wasn't changed in our fork so it's not caused by our changes and by the looks of it its just your code being implemented wrong.

**Comment by tycrek** [comment]
> 
> 
> @tycrek You are passing a java.net.Proxy to the SessionService constructor which doesn't take that as an argument. This wasn't changed in our fork so it's not caused by our changes and by the looks of it its just your code being implemented wrong.

hmm, it worked fine with steveice10's version. With our current code we don't pass any Proxy to a SessionService.

**Comment by tycrek** [comment]
I noticed on my test server it shows *this* failing to connect:

```
com.mojang.authlib.GameProfile@7d096991[id=<null>,name=TheTrueMEG,properties={},legacy=false]
```
perhaps `id` being `null` is the issue

**Comment by tycrek** [comment]
appears to also break Mojang sign-ins, odd. That is fine, I will wait until this PR is merged and ready 🙂

**Comment by rtm516** [comment]
> appears to also break Mojang sign-ins, odd. That is fine, I will wait until this PR is merged and ready 🙂

This also doesn't happen with Geyser. Looking at that stack trace again suggests you are using a heavily outdated or customised version of MCProtocolLib since it says ClientListener.java:55 calls `new SessionService(java.net.Proxy)` which is not the case in the current master version.
https://github.com/Steveice10/MCProtocolLib/blob/master/src/main/java/com/github/steveice10/mc/protocol/ClientListener.java#L55

**Comment by tycrek** [comment]
@rtm516 you are correct, we're using https://github.com/DaMatrix/MCProtocolLib because it's (in theory) intended for the MC server we're interacting with. However just yesterday the admin started a new test server on 1.16.5 so we may need to migrate anyway.

**Comment by WatsOnTV** [comment]
Does this take into account Microsoft 2fa?

**Comment by tycrek** [comment]
> 
> 
> Does this take into account Microsoft 2fa?

It does not (at least, not yet 😉)

**Pull request merged**